zkevm/src/circuit/builder.rs

```diff
@@ -95,125 +103,107 @@ pub fn decode_bytecode(bytecode: &str) -> Result<Vec<u8>, anyhow::Error> {
-                        let caller_code = data.get_code_at(0);
-                        let callee_code = data.get_code_at(1);
-                        trace_code(&mut cdb, caller_code);
-                        trace_code(&mut cdb, callee_code);
-                    }
 
-                    OpcodeId::DELEGATECALL | OpcodeId::STATICCALL => {
-                        let caller_code = data.get_code_at(0);
-                        let callee_code = data.get_code_at(1);
-                        trace_code(&mut cdb, caller_code);
-                        trace_code(&mut cdb, callee_code);
-                    }
-
-                    OpcodeId::CREATE | OpcodeId::CREATE2 => {
-                        let created_code = data.get_code_at(0);
-                        trace_code(&mut cdb, created_code);
-                    }
-                    /*
-                                        OpcodeId::SLOAD | OpcodeId::SSTORE | OpcodeId::SELFBALANCE => {
-                                            let contract_proof = data.get_proof_at(0);
-                                            trace_proof(&mut sdb, contract_proof)
-                                        }
-
-                                        OpcodeId::SELFDESTRUCT => {
-                                            let caller_proof = data.get_proof_at(0);
-                                            let callee_proof = data.get_proof_at(1);
-                                            trace_proof(&mut sdb, caller_proof);
-                                            trace_proof(&mut sdb, callee_proof);
-                                        }
-
-                                        OpcodeId::EXTCODEHASH | OpcodeId::BALANCE => {
-                                            let proof = data.get_proof_at(0);
-                                            trace_proof(&mut sdb, proof)
-                                        }
-                    */
-                    OpcodeId::CODESIZE
-                    | OpcodeId::CODECOPY
-                    | OpcodeId::EXTCODESIZE
-                    | OpcodeId::EXTCODECOPY => {
-                        let code = data.get_code_at(0);
-                        trace_code(&mut cdb, code)
-                    }
+        for (addr, s_map) in storage_trace.storage_proofs.iter() {
+            let (found, acc) = sdb.get_account_mut(addr);
+            if !found {
+                log::error!("missed address in proof field show in storage: {:?}", addr);
+                continue;
+            }
 
-                    _ => {}
+            for (k, val) in s_map {
+                let mut k_buf: [u8; 32] = [0; 32];
+                k.to_big_endian(&mut k_buf[..]);
+                let val_proof: mpt::StorageProof = val.as_slice().try_into()?;
+                let val = verify_proof_leaf(val_proof, &k_buf);
+
+                if val.key.is_some() {
+                    // a valid leaf
+                    acc.storage.insert(*k, *val.data.as_ref());
```

Comment:
seem I need to dig into zkevmi-circuits codes if I want to undertstand more deeply.... will it overwrite if using same `k`?

Reply:
yes. it is by design. zkevm circuits needs the storage before a block and replay all storage write/read in memory. So here we traverse all storage io with reverse order